language/move-prover/interpreter/src/concrete/player.rs

```diff
@@ -0,0 +1,1357 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use num::{ToPrimitive, Zero};
+use std::collections::BTreeMap;
+
+use bytecode::{
+    function_target::FunctionTarget,
+    function_target_pipeline::{FunctionTargetsHolder, FunctionVariant},
+    stackless_bytecode::{
+        AbortAction, AssignKind, BorrowEdge, BorrowNode, Bytecode, Constant, HavocKind, Label,
+        Operation, StrongEdge,
+    },
+};
+use move_binary_format::{errors::PartialVMResult, file_format::CodeOffset};
+use move_core_types::{account_address::AccountAddress, vm_status::StatusCode};
+use move_model::{
+    ast::TempIndex,
+    model::{FunId, FunctionEnv, ModuleId, StructId},
+    ty as MT,
+};
+
+use crate::concrete::{
+    local_state::{AbortInfo, LocalState, TerminationStatus},
+    ty::{
+        convert_model_base_type, convert_model_local_type, convert_model_struct_type, BaseType,
+        Type,
+    },
+    value::{BaseValue, GlobalState, LocalSlot, Pointer, RefTypedValue, TypedValue},
+};
+
+//**************************************************************************************************
+// Internal structs and enums
+//**************************************************************************************************
+
+struct FunctionContext<'env> {
+    holder: &'env FunctionTargetsHolder,
+    target: FunctionTarget<'env>,
+    ty_args: Vec<BaseType>,
+    label_offsets: BTreeMap<Label, CodeOffset>,
+}
+
+impl<'env> FunctionContext<'env> {
+    pub fn new(
+        holder: &'env FunctionTargetsHolder,
+        target: FunctionTarget<'env>,
+        ty_args: Vec<BaseType>,
+    ) -> Self {
+        let label_offsets = Bytecode::label_offsets(target.get_bytecode());
+        Self {
+            holder,
+            target,
+            ty_args,
+            label_offsets,
+        }
+    }
+
+    pub fn code_offset_by_label(&self, label: Label) -> CodeOffset {
+        return *self.label_offsets.get(&label).unwrap();
+    }
+}
+
+//**************************************************************************************************
+// Backbone
+//**************************************************************************************************
+
+/// Entrypoint of the step-by-step interpretation logic
+pub fn entrypoint(
+    holder: &FunctionTargetsHolder,
+    target: FunctionTarget,
+    ty_args: &[BaseType],
+    typed_args: Vec<RefTypedValue>,
+    global_state: &mut GlobalState,
+) -> PartialVMResult<Vec<TypedValue>> {
+    let ctxt = FunctionContext::new(holder, target, ty_args.to_vec());
+    let local_state = exec_function(&ctxt, typed_args, global_state)?;
```

Comment:
It's not really step-by-step on bytecode level but on function invocation level?

Reply:
> It's not really step-by-step on bytecode level but on function invocation level?

Seems not step-by-step at all for now. We can consider this good to go for the first version, but eventually, we may want an executor which can really run step-by-step.

Reply:
The `exec_bytecode()` is the `step-by-step` part, or more precisely, `bytecode-by-bytecode`.

Reply:
A real step-by-step execution would provide a public entrypoint `fn exec_next_step(&mut State)` or similar which would execute the next bytecode instruction and then return. This would allow single-stepping through execution and other advanced instrumentation scenarios. I'm not saying we need this now, just confused because you mentioned stepping. If I do not misunderstand, what you do right now is not single-stepping in this sense.

Reply:
Ah, I see. But I still feel that the `exec_bytecode` function is implemented with this spirit.
```
fn exec_bytecode(
    ctxt: &FunctionContext,
    bytecode: &Bytecode,
    local_state: &mut LocalState,
    global_state: &mut GlobalState,
) -> PartialVMResult<()> {}
```
It takes a bytecode, and two mutable states (local and global) and just execute that bytecode and then return.